### PR TITLE
Add authentication module

### PR DESCRIPTION
**Pull request by lucian-v**

### CODE REVIEW

File: template/src/modules/@auth/hooks/useAuth.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    authState: '',
+  })
+
+  const handleStateChange = useCallback((authState: ValidAuthState, authData?: AuthData) => {
+    if (authState === 'signedOut') {
+      authState = 'signIn'
+    }
+
+    setState(state => {
+      logger.debug(`state changed '${state.authState}' -> '${authState}'`, authData)
```
Comment (danutDiac) [inline template/src/modules/@auth/hooks/useAuth.ts:22]: Can we remove this?

File: template/src/modules/@auth/hooks/useConfirmSignUp.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    } catch (error) {
+      logger.error(error)
+      throw error
+    }
+  }
+
+  const resend = async (): Promise<void> => {
+    try {
+      await Auth.resendSignUp(username)
+      logger.debug('code resent')
```
Comment (danutDiac) [inline template/src/modules/@auth/hooks/useConfirmSignUp.ts:25]: How about this one?

File: template/src/modules/@auth/hooks/useSignIn.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    navigate(postSignInRoute)
+  }
+
+  return async (username: string, password: string): Promise<void> => {
+    try {
+      const user = await Auth.signIn({
+        username,
+        password,
+      })
+      logger.debug(user)
```
Comment (danutDiac) [inline template/src/modules/@auth/hooks/useSignIn.ts:28]: Here

File: template/src/modules/@auth/hooks/useAuth.ts (hunk 1 above)
Comment (lucian-v) [reply template/src/modules/@auth/hooks/useAuth.ts:22]: I would actually keep them, as `logger.debug` output doesn't appear in the console, unless you set your log level to `debug`. 
What do you think?

File: template/src/modules/@auth/ProtectedRoute.tsx [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+}: ProtectedRouteProps) => {
+  const { authState } = useAuthContext()
+  if (authState === '') {
+    return null
+  }
+
+  if (authState !== 'signedIn') {
+    return <Redirect to='/auth' noThrow />
+  }
+  // TODO check group/role based permissions
```
[the remark this answers is not in the document]
Comment (lucian-v) [reply template/src/modules/@auth/ProtectedRoute.tsx:23]: Added

File: template/src/modules/@auth/hooks/useAuth.ts (hunk 1 above)
Comment (danutDiac) [reply template/src/modules/@auth/hooks/useAuth.ts:22]: I guess it's less painful that it doesn't appear in the console.
My thoughts were that they served their purpose, and I would "clean" by removing them.

**Pull request merged**